### ISSUE TITLE
[UX] Add Image Library modal to Image fields

### ISSUE THREAD

**Issue by laryn** [backdrop/backdrop-issues#3297]
## Describe your issue or idea

I was adding the [filefield_sources error](https://github.com/backdrop-contrib/filefield_sources/issues/10) to my (long) rainy day list and then had a different thought. How hard would it be to add an option/link to core image fields that pulls up our [shiny new media library view](https://github.com/backdrop/backdrop-issues/issues/3134) in a modal? That would be much better than the Filefield Sources solution, I think.

So you could upload a new file (as usual), or click a link to browse the library and attach an existing image to the file field you're using.

Weekly update?
Latest updates on 4th December 2019.
A new PR is under development and test at https://github.com/backdrop/backdrop/pull/3011. At the moment it is only partially complete but it does open dialog in a modal.
More updates to this new PR today, 18th December. Still not complete.
Advocate: @laryn 

---

~~PR by @Graham-72: https://github.com/backdrop/backdrop/pull/3011~~
PR by @quicksketch (based on @Graham-72): https://github.com/backdrop/backdrop/pull/3028

**Comment by laryn** [comment]
Rough image mockups for discussion (no code yet).

![upload-image-1](https://user-images.githubusercontent.com/871421/45895239-48dd3200-bd96-11e8-858a-0b5b42a4f798.jpg)

![upload-image-2](https://user-images.githubusercontent.com/871421/45895242-4bd82280-bd96-11e8-8afb-c975999837de.jpg)

**Comment by olafgrabienski** [comment]
I also like the idea very much!

@laryn - For reference I'd suggest to also have a look at #1307 (part of the meta issue #1448) which sounds quite similar. I'm not sure if it makes sense to continue the work there, or if it's better to restart with a more simple approach which makes no claim to be complete.

**Comment by klonos** [comment]
I like this idea a lot as well 👍

**Comment by korontari** [comment]
I like this idea very much and would be glad to see it implemented as soon as possible. This issue in one of my major inconveniences when using Backdrop.

**Comment by jenlampton** [comment]
I agree, this looks fantastic, and if its easier to start from this angle (rather than https://github.com/backdrop/backdrop-issues/issues/1307) I'm all in favor of it. Progress is progress!

**Comment by quicksketch** [comment]
The code for the toggle between different file selections is very specific to the image dialog right now (I believe the code for it is also in filter.module). I'd like to see that code refactored into something generic and made a part of File module (essentially FileField Sources) if we're going to use it in multiple places. But that is a lot of work. Maybe they don't need to be the same thing? I've definitely been bitten by the "hey we have _almost_ the same thing in two places" bug only to have the refactoring create something much messier than just having two different implementations.

**Comment by klonos** [comment]
@quicksketch I don't see how these can be different things. Both the editor toolbar button and the "Choose file" button should allow either uploading new, or selecting existing files. It's just that the one is to insert the image to the file field, while the other is to also generate the required HTML to insert it in the body/text field.

Having said that, I don't know the complexity behind actualising this. I would need to see a PR to understand if the code we end up with is spaghetti or not.

**Comment by klonos** [comment]
We should allow image styles to be selected from the dialog as well (#3589), which will basically be https://www.drupal.org/project/iss

**Comment by Graham-72** [comment]
@laryn I have started to work on this, adapting code from the Image Library implementation in the filter module, with which I have some experience. I am working on adapting the necessary javascript at the moment. I aim to have a proof of concept PR by end of October 2019. I am away on holiday next week so cannot do more until after 7th October.

**Comment by laryn** [comment]
@Graham-72 Sounds amazing -- thanks again!

**Comment by docwilmot** [comment]
We also need to update the the image library to allow inserting other media.

**Comment by klonos** [comment]
Yeah, and perhaps start calling it **Media** Library instead 😉

**Comment by Graham-72** [comment]
@laryn Initial PR now published at https://github.com/backdrop/backdrop/pull/2953.

**Comment by vstemen** [comment]
Cool.  I'm happy to see some work being done on this issue. :-)

**Comment by laryn** [comment]
Great, thanks @Graham-72 -- I'll pull out some screenshots and a few comments here. Anyone else feel free to chime in, too.

**1) Field display when adding a new piece of content with an image field:**

> ![Screen Shot 2019-10-28 at 10 46 35 AM](https://user-images.githubusercontent.com/871421/67693814-563cf680-f970-11e9-82ae-28ae818760bc.jpg)

Comments:

 - There are some visual spacing/sizing issues that I'm assuming can be dealt with closer to the end of the process, if still present

**2) Field display after clicking "Select from library":**

> ![Screen Shot 2019-10-28 at 10 49 20 AM](https://user-images.githubusercontent.com/871421/67693976-9d2aec00-f970-11e9-8621-6e89a1cabc46.jpg)

Comments:

 - Would it be possible to have this pop up in a modal? Then the width could be larger for bigger libraries, and if the user decides the image they want isn't present in the library they just close the modal and have the upload field automatically in front of them again.
 - The left side "File properties" displays no information currently, and clicking a file does not fill that information out but rather jumps to a "Confirm image" display that removes the "File properties" completely. (Aside: just noticed that hovering over an image does display a few image details in the "File properties" pane, although instructions say to click, not hover)

**3) Confirming selection of an image from the library:**

> ![Screen Shot 2019-10-28 at 10 55 08 AM](https://user-images.githubusercontent.com/871421/67694504-6e614580-f971-11e9-9f8f-63d06455abfb.jpg)

Comments:

 - Is the "File properties" pane intended to be shown here?
 - The "Image file" field should probably be locked so you can't type in it over top of the filename
 - There is a "Confirm" button, but no "Back" or "Cancel" buttons

**4) After confirming an image: **

> ![Screen Shot 2019-10-28 at 10 57 20 AM](https://user-images.githubusercontent.com/871421/67694680-c1d39380-f971-11e9-8a2a-0555c47a2662.jpg)

Comments:

 - Again, it's possible to type in the "Image file" field (is this field required to show, or should it be hidden/removed?)
 - Image width and height are tacked on here, maybe intended to be moved into a "File properties" pane when implemented?
 - May be beyond scope, but should we have a "Manage image" link on image fields that have an image loaded, which goes to the **Content > Manage files** page for that particular image, where you can edit/replace the name/description of the file and update the file without changing the filename (and other future functions, presumably).

@Graham-72 – would it help to have graphical mockups to work from, or do you have a pretty good sense of direction you'd like to go next for these screens?

**Comment by Graham-72** [comment]
@laryn thanks very much for your comments. I will be making some modifications soon. As you suggest I think we can leave final styling until we have settled some other issues.

**Comment by Graham-72** [comment]
I think we ought to provide for setting of the alt and title fields as part of the process, something I don't think we can do at present?

**Comment by jenlampton** [comment]
The Alt and Title from the image field, or the Alt and Title fields from the File entity?

**Comment by Graham-72** [comment]
Yes, an important point. If we reuse an image by selecting it from the image library, should we also be reusing its Alt and Title field settings, or should these be set for each use?

**Comment by laryn** [comment]
It would be great if the alt/title values set when uploading an image were set as global defaults for that image and loaded by default on additional uses (but could be overridden at that point in a per instance basis)

**Comment by Graham-72** [comment]
Also must take note of issue #4007

**Comment by Graham-72** [comment]
I have made some changes to the PR and have more in the pipeline.
I would like the screen 'Confirm image' to show after the file upload in the same way that it does after selection by clicking in the image library, and then to give the opportunity to enter alternate and/or title text.

**Comment by jenlampton** [comment]
Today in the weekly meeting @quicksketch asked if the image gallery was opening in a modal, and it looks like it isn't right now. Is that part of the long-term plan?

**Comment by Graham-72** [comment]
I started by thinking it would be in a modal but now I'm not sure whether that is necessary or what the benefit would be, but yes, to be considered.

**Comment by jenlampton** [comment]
I also am not sure a modal would be better UX than what we have now. A
modal makes more sense in the editor.

I would like to see the thumbnail view behind an Ajax call so that it's not
loaded or rendered on an edit form if it's not needed.

The results of the view will also change for the second image field when an
image is added into the first one, and rendering the thumbnail view too
early would miss that. If you wanted to add an image once but use it twice,
you'd need to save the form and edit again to see the new image in the
thumbnail view.

**Comment by docwilmot** [comment]
I also feel this would be better in a modal. Better UX, especially if you have a lot of images.

**Comment by Graham-72** [comment]
I have made some revisions to the PR but am now unlikely to do anything for a few days because I will be away from base. My next intention is to get the alt and title fields working. Other things later - step by step approach.

**Comment by Graham-72** [comment]
> Better UX, especially if you have a lot of images.

I will need convincing on this.

**Comment by olafgrabienski** [comment]
Regarding modal dialog or not, at the current state of the PR it's hard to evaluate different interface approaches. Let's however keep the question in mind! In the meantime we could have a look at how Drupal 8 and WordPress manage their image / media libraries.

**Comment by Graham-72** [comment]
I have made more revisions to the PR. The sandbox demonstrates that it is working successfully though there are a large number of test failures which I will now investigate.

The alt and title fields are usable though need to be reviewed along with issue #4007 and I have made no attempt to store 'initial' values that are re-usable when re-using an image from the library. Where would these values be stored?

I have included my personal preference of showing additional information about each image in the library because of situations where the same image appears more than once in the library, perhaps having a different size - see issue #3386 for examples. To include this in full I had to import into the sandbox a revised copy of the image library view config settings.

**Comment by laryn** [comment]
Current draft walkthrough/snapshot:
![image-library-191113dev](https://user-images.githubusercontent.com/871421/68813280-f8c8cb00-063a-11ea-8587-14e39aa70f34.gif)

**Comment by Graham-72** [comment]
@laryn thanks for the excellent walkthrough-screenshot.

**Comment by docwilmot** [comment]
> > Better UX, especially if you have a lot of images.
> 
> I will need convincing on this.

both aesthetic and practical. If you have a gallery site with 500 images this will really clutter the `node/add` form. Furthermore it seems that all the thumbnails are loaded on page load, but hidden until you click the library toggle: with many images you will be loading every single one each time. Even worse if you have more than one image field.

(unless there is a plan for some AJAX paging)

Plus we may need more actions on this now or future. How about searching for an image, or sort by size etc? Or selecting an image style? I feel like selecting and sorting and searching for images is an action that should be best performed outside the node edit flow.

Maybe personal preference, but I think a modal is best.

**Comment by Graham-72** [comment]
Yes, it should be paging the images as it does for us in CKeditor but I guess this needs to be in a modal to work. At the moment if you have more than 20 images (value as preset in the view) the 'next>' button will not cause the next page to appear.

We have issue #3293 about searching within the image library. This is marked as 'good first issue' but my attempts have not succeeded so far. It would be good to provide this.

**Comment by laryn** [comment]
Update from @Graham-72 for meetings today: At this point "...opening a modal is not just preferable but essential in order to get the paging of the Image Library to work. I have also been looking into adding a search on filename field - issue #3293 but have hit a blockage. Regarding the modal, it seems I am going to have to open it from within the javascript because of the constraint of building on the basic file upload dialog. The ckeditor.js javascript file would seem to do this so there is something I can adapt, but it is non-trivial!"

**Comment by laryn** [comment]
@Graham-72 has made progress on an alternate PR which opens in a modal. Functionality is not all in place yet and [I've posted some preliminary comments on the PR itself.](https://github.com/backdrop/backdrop/pull/3011#issuecomment-561758436) 

He has added a filename filter to the core Image Library view, and I've added exposed sort for date and name, for discussion. The revised View is attached here as a text file in case anyone wants to import it. ([Exposing sort fields is broken](https://github.com/backdrop/backdrop-issues/issues/1324) and takes a few extra steps to get to work so importing the view on a local may be quicker than recreating). This works in the new modal popup he's working on, but breaks functionality in the WYSIWYG image modal for some reason.

[ImageLibraryView-ExposedFilters.txt](https://github.com/backdrop/backdrop-issues/files/3927883/ImageLibraryView-ExposedFilters.txt)

Here is a quick demo walkthrough of the current status:
![ImageFieldLibrary-120519-status-update](https://user-images.githubusercontent.com/871421/70252563-41aa0600-1747-11ea-9ee0-fe7d4786a854.gif)

**Comment by laryn** [comment]
Latest status here:

> Things proved more difficult than I expected. I still have to arrange for closure of the modals with necessary updating of fields. Some of these things are very hard to find out how to do!
> You may not be keen on the revised UI and I am not sure how it might be improved without more knowledge and effort.
> And we still have to tackle the tests, presumably adding some.

I took a look at the latest PR adjustments, which add an auto-complete field to the mix:
![Screen Shot 2019-12-19 at 11 56 27 AM](https://user-images.githubusercontent.com/871421/71197151-9cb71f00-2256-11ea-801a-a4ca90d618a0.jpg)

I suggested perhaps using Filefield Sources as guidance on the UI here rather than having everything showing:

![Screen Shot 2019-12-19 at 11 49 02 AM](https://user-images.githubusercontent.com/871421/71197203-b3f60c80-2256-11ea-999a-672b14be332e.jpg)

Also I posed the question of whether the autocomplete should be discussed as a separate/later addition to focus on the image library piece and see if it's possible to get that into 1.15.

Help/direction on the tests will be desirable when we get to that point. 
Any other comments from others at this point?

**Comment by klonos** [comment]
Sorry for coming late to the party. I had a high-level play with the current implementation in the PR sandbox, and my first reactions were the following:

- The image form widget is now occupying a huge space.
- Why not use a single "add image" button, which pops open a dialog, in which all controls (upload/reference/browse) are available as tabs? This would make the UX consistent with the "insert image" button in the rich text editor.

The general idea being to reuse UI patterns, and to minimize the space the add image form widget takes, by moving everything into a dialog. The widget should be reduced to a single "Add image" button, an image thumbnail (if an image is selected), and the help text.

Other feedback:

- I like the addition of the image info/details, but it could use some styling.
- I could not actually select an image.
- The styling of the exposed filters in the "Select from Image Library" dialog differ from the "Insert image" dialog (the "name" filter and the "apply" button are smaller)
- While on the "Insert image" dialog, the exposed filters have no gray border around them, but soon as you click the next/previous buttons on a paginated view, the border shows thereafter.
- The "Select from Image Library" dialog has no borders in its exposed filters whatsoever.
- UX: the insert button should be sticky to the bottom of the dialog.

**Comment by jenlampton** [comment]
@klonos how many of these are blockers, and how many can be follow-ups?

**Comment by klonos** [comment]
@jenlampton I see the following as blockers:

- UX: the insert button should be sticky to the bottom of the dialog.
- Cannot actually select an image.
- Fix the exposed filters functionality (if not, then remove and follow-up)
- Styling/CSS issues.

Can be follow-ups:

- Image info/details can be added later (will reduce code/scope).
- Autocomplete can be a separate issue.
- Exposed filters (if we cannot fix by 1.15 feature freeze)

The bit where we consolidate everything into a single dialog that opens via a single "add image" button is also a biggie I imagine. Consider multi-value fields, and how huge the forms will end up being.

**Comment by klonos** [comment]
BTW, rthis seems to be a duplicate: #1307

Note to self: copy any useful comments from that issue, and then close.

**Comment by quicksketch** [comment]
> Image info/details can be added later (will reduce code/scope).

Agreed, I found this behavior to be confusing when no images are available and/or no image had been selected yet.

![image](https://user-images.githubusercontent.com/100206/71489203-3fcce300-27d9-11ea-8bb9-063a7f4c7159.png)

The dependency on hover also may not work well on mobile devices.

> Autocomplete can be a separate issue.

:+1: I think it makes more sense to have the filtering occur after the dialog has been opened (via exposed filters) than to put an auto-complete on the node form itself.

---

I read over the PR.

Instead of extending the form in `image_field_widget_process()`, I think it would be preferable to modify the `file_managed_file_process()` function to add new form API properties. We could make it so that something like this would be acceptable:

```
  $form['image'] = array(
    '#type' => 'managed_file',
    '#upload_location' => 'public://example',
    '#browser_view' => 'image_library',
  );
```
Doing this would allow us to style the file upload element all-together, sort of similar to how FileField Sources works (which also extends the "managed_file" element).

For the UI, I think if we can reduce the visual clutter to something as simple as this:

![image](https://user-images.githubusercontent.com/100206/71490056-b7047600-27dd-11ea-80a7-5e24849a13ba.png)

As @Graham-72 already has found, we have to use a button element to open the dialog because it has to be submitted via POST to send the form data. So to make that UI work we'll actually still use a button like we are currently, but hide it with JavaScript.

---

I know that's a lot to ask at this point. I'll try a hand at this over the next few days and see if it's feasible.

**Comment by Graham-72** [comment]
@quicksketch thanks for your comments. I have just added my latest changes to the PR and things now look somewhat different.

I have not managed to complete the modals and have been going round in circles for many days trying to understand and apply the limited Backdrop documentation on these. I have an ambition to master them and write a tutorial with full examples of how to save, close or cancel them!

>  modify the file_managed_file_process() function to add new form API properties
 
This sounds a great idea and should make things easier. Thanks. 

@laryn please have a look at the revised UI but you will need to be sure your browsers downloads the revised `file.js`. I would go for a new sandbox but this would mean having to upload a new set of images.

**Comment by quicksketch** [comment]
@Graham-72 I posted a new PR at https://github.com/backdrop/backdrop/pull/3028 that takes your work and continues to move things around:

- I moved almost everything out of image module and into file module. The `managed_file` element now has a generic ability to use a file browser based on a view through setting a `$file_element['#browser_view']` property.
- However, as we only currently ship with a view suitable for browsing images, the only field that utilizes this feature is the image field. *You must now enable the browser* on the image widget instance settings, such as on `/admin/structure/types/manage/post/fields/field_image`:
![image](https://user-images.githubusercontent.com/100206/71534755-cce56a00-28b5-11ea-9a3a-fbc6e040b9a8.png)
- I removed the various ways of selecting a file to include only the browser capability. The UI for images on the node form now currently looks like this (after enabling the browser):
![image](https://user-images.githubusercontent.com/100206/71534793-103fd880-28b6-11ea-8472-3459740e60ab.png)
As I mentioned before, I think a text link would be preferable to understate this and reduce the button clutter, but that will take a small amount of extra JS to hide the button and then trigger it when the link is clicked.

I had to wrack my brain a bit to figure out how to get dialogs to pass information to a different part of the form. I don't think we have any examples in core of a situation like this. CKEditor's file dialog is all client-side JS. And modules like Layout and Views are constantly updating a partially-saved object in the `state` table. This situation is server-side based, but we don't have the object store to update. So the implementation looks sort of like a mix between a CKEditor modal and a Views modal.

To set the selected image in the node form, I used server-side callbacks to pass first a selector for the hidden FID element, and then to pass back the selected FID. The modal `dialog:afterclose` event is then used to set the updated FID, and then it clicks the hidden "Upload" button to trigger updating the form. This is the same trick FileField Sources uses to set files, that it just updates the hidden FID field and then submits the upload normally. File module will automatically pick up any manually set FID and uses it as though it were an uploaded file.

This PR is still doing a bit too much. The file preview information is image-specific and wouldn't work for generic file selection. I think the closer we can get this to match the experience of the CKEditor image dialog, the better. So as we don't have that in CKEditor, I don't think we should be adding it here either.

@Graham-72, I'd love your feedback. As there is still work I know needs done I'll iterate on this a few times. I'm using your initial dialog code but then the form submission approaches from FileField Sources. If you have implementation questions, post them as comments on the PR and I'll answer them there.

**Comment by klonos** [comment]
I really like where this is heading implementation-wise, working towards making this a generic file browser 👍 (as opposed to image-specific).

I could not test the actual browser dialog, because as I mentioned in the PR, I could not get it to open. Nevertheless, I have the following feedback:

1. The browser should also be available when selecting default images/files for fields:

    ![Screen Shot 2019-12-29 at 7 33 06 am](https://user-images.githubusercontent.com/2423362/71549264-e1894580-2a0e-11ea-9268-f9a4bd4aeea5.png)

2. In the screenshot provided by @quicksketch in the previous comment, the label of the standard upload widget is "Browse...", whereas in the PR sandbox it is "Choose file". IMO, this does not differentiate it enough from the "Select existing file" label of the newly-introduced button of the file browser:

    ![Screen Shot 2019-12-29 at 7 48 32 am](https://user-images.githubusercontent.com/2423362/71549300-a63b4680-2a0f-11ea-9717-0d925978a3eb.png)

    ...I understand that "existing" is meant/implied here in the context of the site/CMS, but this seems like a point that could potentially confuse users. I have tried a few options, but do not have any specific alternative ATM, as everything I came up with was too wordy 😞. This can be a UX follow-up though.

3. I have tried adding a second image field in the same form (allowing unlimited values, if that matters), and have noticed that when clicking the "Select existing file" button of the second field, the "Choose file" button of the first field was momentarily disabled (whereas I would expect the matching one to do so):

    ![Screen Shot 2019-12-29 at 7 35 01 am](https://user-images.githubusercontent.com/2423362/71549436-51003480-2a11-11ea-8c2f-c0c557c6f0c1.png)

   ...since as I mentioned, the dialog would not open for me, I could not test whether the actual file upload goes to correct field or not.

**Comment by vstemen** [comment]
> in the PR sandbox it is "Choose file". IMO, this does not differentiate 
> it enough from the "Select existing file" label of the newly-introduced 
> button of the file browser:

What about something like "Select from library"?

**Comment by quicksketch** [comment]
> the label of the standard upload widget is "Browse...", whereas in the PR sandbox it is "Choose file".

The "Browse..." vs. "Choose file" wording is browser-specific. Unfortunately file upload elements do not let you specify the label of the button. While you can hide the entire element and make a replacement, the label isn't something we changed in this PR.

> the "Choose file" button of the first field was momentarily disabled (whereas I would expect the matching one to do so)

Ah! Great catch! So a work-around File module includes is that it disables upload fields during AJAX requests to prevent files from being uploaded when unrelated buttons are clicked. For example files aren't normally uploaded when you click "Add more" on a _different_ field. However the selector makes an exception for the "Upload" button, which clearly should include the selected file. The "Select Existing File" button is close enough to that selector that it also has the same behavior. You're absolutely right that we should disable the file upload when that button is clicked. This entire problem of selecting but not actually uploading is fairly uncommon at this point though, since file uploads now auto-upload immediately after selecting a file. That was more of an issue in D7 when you had to explicitly click "Upload" after selecting a file.

@olafgrabienski noted:

> I've started to test the PR in the sandbox site and could open the modal only once, then it didn't open anymore. When I click the Select existing file button, the cursor jumps to the Title input of the node form.

I think this is a larger issue that happens already with all buttons on the node form. It seems some recent changes to Firefox made it so that the browser jumps to the field that has focus when any button is clicked. You can suppress this behavior by first focusing the button with the tab key, then clicking it. As I think this happens with all buttons, not just the browser button, we may want to track that separately.

**Comment by quicksketch** [comment]
PR https://github.com/backdrop/backdrop/pull/3028 updated with the following:

- Removed the "Image Details" portion of the dialog so-as to match the CKEditor dialog and simplify the UI.
- The file browser dialog now has a cancel link.
- Fixed the focus state of the current selected image in the dialog.
- Fixed a lot of double event binding issues. That may fix @klonos's problem of the dialog not opening at all.

**Comment by Graham-72** [comment]
@quicksketch here's my initial feedback. I'm glad and somewhat relieved to see that you have produced a new PR to adopt (hopefully) for release. My PR is very much a feasibility study to investigate what might be done and what is might look like. Also I'm reassured by your explanation of the difficulty of getting information from a modal dialog to another part of the form. I was thinking that there must be an obvious solution that my 77 year-old brain was failing to grasp! I shall slowly study your PR with great interest.

I have two points regarding usability. I believe this is particularly important for features like this which are likely to be used by a class of user I would call 'office user' rather than site builder, i.e. someone with a need to edit page content, who can use a wordprocessor, but has no real knowledge of how websites work or how to use Backdrop. We know that 'choose file' relates to finding a file on our PC to upload to our website but they may not know this. I believe it is worth having a little bit more information on the screen to guide them, in this case the subheading 'Upload image' would seem to be an adequate clue, like this
![image (1)](https://user-images.githubusercontent.com/10559274/71559207-6dbd5c00-2a53-11ea-8746-baa8e2930078.png)

My other point about usability is that I am sad that you are not including image details. When one is working with a website with 100+ images, some of which are different sized versions of the same image (for legitimate reasons) it is useful to know more about each file and to have that information available just when making the decision. I know we don't have it in our CKeditor version and I hope we might try it here and update the CKeditor version if reaction is favourable. I think there might be other refinements we could make from experience. Though I agree with the aim of keeping them similar.

I would like to keep my PR active for a little longer in order to deal with any problems, until such time as the new one is merged and provides a base for further work.

**Comment by olafgrabienski** [comment]
@quicksketch Thanks for updating thePR! Still, when I click the "Select existing file" button in the PR's sandbox, the dialog doesn't open -- not only in Firefox but also in Chrome, Opera and Safari. (Maybe only a sandbox issue? According to https://github.com/backdrop/backdrop/pull/3028#issuecomment-569545468 it works better in a custom test environment.)

**Comment by quicksketch** [comment]
Thanks @Graham-72 for your feedback! I do feel terrible that I've removed functionality you worked hard to put together. I think the image details is a good idea, and it provides valuable insight into the images that are selected. However in the interest of time and consistency, I didn't think it practical to try and complete it for 1.15. I definitely think we should use the previous PR as a reference for adding that feature later (note even if we close it, the Diff and branch are actually maintained here on GitHub).

----

I updated the PR at https://github.com/backdrop/backdrop/pull/3028 with a few more changes:

- Fixed the issue where the modal would not open at all. This was a problem caused by any multiple value file or image field on the form. I didn't see this on my localhost because I was only testing a single image field.
- Fixed the temporary disabling of upload fields when the "Select existing file" button is clicked. Now all file upload fields (including the one next to the button) are disabled for 1 second while the dialog is opened. This prevents accidentally uploading a file when you intend to use the image library.

**Comment by olafgrabienski** [comment]
> Fixed the issue where the modal would not open at all. This was a problem caused by any multiple value file or image field on the form. I didn't see this on my localhost because I was only testing a single image field.

My apoloigies, @quicksketch! I guess I've changed the old PR sandbox to multiple images without mentioning it. I'm glad you found the reason anyway!

Now, I've tested the fix in the current sandbox website. Works both with single images (cf. *Posts*) and multiple images (cf. new *Event* content type).

One side note: In the *Event* content type where multiple images are allowed, I tried to select more than one image in the image browser but that's not possible. I had to select one image, insert it, open the image browser again, select the next image and so on. In a way, this behavior is consistent to how images are uploaded where it's also not possible to select multiple images in one step without additional modules. It felt however a bit more strange and kind of outdated in the image browser. Without thinking about it, I'd expected to be able to select more than one image in the same step. Maybe a follow-up issue?

**Comment by quicksketch** [comment]
> Without thinking about it, I'd expected to be able to select more than one image in the same step. Maybe a follow-up issue?

We'll need the ability for the "managed_file" element to handle multiple files before we can make this work for the image browser as well. We have an existing issue for that prerequisite at #440, and a similar issue about JS-library solutions at #1380. I agree selecting multiple images/files would be handy, though it'll take some additional work for representing how to select/deselect images and deal with multiple pages of files in the view.

**Comment by Graham-72** [comment]
It seems to be working well. I have done some testing with multiple images and two image fields in a node - no problems, great! The paging in the image library is working too. Thanks @quicksketch .

**Comment by herbdool** [comment]
Looks like it's coming along. Don't know if it's ready for review but I saw a commented out line in the CSS.

By the way, what happens if someone disables the view?

**Comment by olafgrabienski** [comment]
The issue is labeled as `needs - more feedback` and `pr - needs work`, are the labels still adequate?

**Comment by quicksketch** [comment]
> Don't know if it's ready for review but I saw a commented out line in the CSS.

I removed the CSS changes since those were in place to facilitate the file details sidebar (which is also removed). So now the PR is pretty small and gets the job done. Updated the status to indicate it's ready for review.

> By the way, what happens if someone disables the view?

Looks like in both this dialog and in the CKEditor dialog, it just leaves a blank white space in the dialog. We should provide a message indicating why the space is empty with instructions to re-enable the view. At least the "image_library" view is locked so that it can't be deleted, but you're right it can be disabled.

**Comment by jenlampton** [comment]
Changing labels bc @herbdool said "works for me" on the PR: https://github.com/backdrop/backdrop/pull/3028#issuecomment-570012489

**Comment by jenlampton** [comment]
I left a comment on the PR, but it looks like usage records are not being updated for re-used images. That may need to be fixed before this gets merged.

**Comment by jenlampton** [comment]
I did a quick code review and it looks great. I had 2 questions, one a minor one line cleanup, and the other a question about the `button` type vs `submit` that may be irrelevant :) LGTM!

**Comment by jenlampton** [comment]
Feedback on PR resolved, RTBC!

**Comment by quicksketch** [comment]
I made a few additional changes:

- If the "image_library" view is disabled, a message is now provided instead of a blank dialog. If the user has permission to modify views, a link to the Views Administration is provided.
![image](https://user-images.githubusercontent.com/100206/71647566-dacc2e00-2cad-11ea-9cd9-e3f8f2dec5ce.png)
- The button to browse existing images is now hidden if JS is disabled, since the option is dependent on JS to work.
- I added a sort to the "image_library" view so it is now reverse-chonological (newest images first).

**Comment by jenlampton** [comment]
@quicksketch I found 2 text changes needed in the latest PR: https://github.com/backdrop/backdrop/pull/3028#pullrequestreview-337510807

**Comment by herbdool** [comment]
I took a look at the final code changes. It looks good to me.

**Comment by quicksketch** [comment]
I've merged https://github.com/backdrop/backdrop/pull/3028 into 1.x for 1.15.0. Thanks everyone!!
This was quite the group effort! Extra thanks to @Graham-72 for undertaking this early and @laryn for advocating for this particular issue and providing regular status updates and testing.
